### ISSUE TITLE
Add pagination to the list of roles

### ISSUE THREAD

**Issue by helfi92** [taskcluster/taskcluster-web#307]
If you're looking for both a backend and a frontend change, this task would be ideal for you. Right now if you navigate to https://taskcluster-web.netlify.com/auth/roles, you will notice the full list of roles is displayed. We want to paginate this list so that we can render it similar to the [clients view](https://taskcluster-web.netlify.com/auth/clients). In order to do that, we'll first need to do a backend then a frontend change.

#### Backend

The orginal bug filed for this task can be found [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1455130). You can create a bugzilla account and comment there to continue the discussion. The change will most likely happen in the [`taskcluster-auth`](https://github.com/taskcluster/taskcluster-auth) repository.

#### Frontend

Once the backend work is done, you'll want to use the [`ConnectionDataTable`](https://github.com/taskcluster/taskcluster-web/tree/master/src/components/ConnectionDataTable) component to render the table. The change will happen in the taskcluster-web repository.

PS: If you just want to tackle the backend part, that's okay too. Someone else can hook up the frontend later.

**Comment by helfi92** [comment]
👋 @OjaswinM, definitely! For the backend part of this issue, do you mind commenting on https://bugzilla.mozilla.org/show_bug.cgi?id=1455130 to have the bug assigned to you?

**Comment by OjaswinM** [comment]
Alright, I'll get back to you if i need any help

**Comment by helfi92** [comment]
@OjaswinM You should be ready to go now. For the graphql file, you can probably just use https://github.com/taskcluster/taskcluster-web-server/pull/71#issuecomment-449107612. We use pagination on many of our views so hopefully it will make it easier to understand. Let me know if you have questions. Can't wait to see what you come up with :)

**Comment by OjaswinM** [comment]
Hey actually I have my finals coming and would be busy for 2 to 3 weeks. I would let someone else pick it up from here. I would love to work on it if it's still open when I'm free again :)

**Comment by helfi92** [comment]
Sounds good, thanks for the heads up @OjaswinM. We'll see where we when you're back. Good luck with your finals :)

**Comment by arshadkazmi42** [comment]
@helfi92 I would like to work on  this.
So does this needs backend change also? Or backend change is implemented?

**Comment by helfi92** [comment]
The backend change has been implemented :) go for it

**Comment by arshadkazmi42** [comment]
Do I need to setup backend in my local for this? Or there is some backend url which I can use during development?

**Comment by helfi92** [comment]
Yes, please. [`taskcluster-web-server`](https://github.com/taskcluster/taskcluster-web-server) is what you need. I believe only the root url is needed to have the service up and running locally. See [here](https://github.com/taskcluster/taskcluster-web-server#running-taskcluster-web-server-locally) for instructions.

**Comment by arshadkazmi42** [comment]
I started with the setup, getting some issues.
Where can I get these `TASKCLUSTER_ROOT_URL`, `TASKCLUSTER_CLIENT_ID`, `TASKCLUSTER_ACCESS_TOKEN`

**Comment by helfi92** [comment]
> Where can I get these TASKCLUSTER_ROOT_URL, TASKCLUSTER_CLIENT_ID, TASKCLUSTER_ACCESS_TOKE

https://github.com/taskcluster/taskcluster-web-server#taskcluster-credentials

**Comment by arshadkazmi42** [comment]
On running `eval $(./taskcluster signin --name taskcluster-web-server)`
its stuck at
```
Listening for a callback on: http://localhost:43483
Opening URL: https://tools.taskcluster.net/auth/clients/new?name=taskcluster-web-server&description=Temporary+client+for+use+on+the+command+line&scope=%2A&expires=1d&callback_url=http%3A%2F%2Flocalhost%3A43483
```
I started my `taskcluser-web` using yarn start, its giving network error at `auth/roles`

**Comment by helfi92** [comment]
Running the eval command will open a web page. The web page should be used to sign in. Once you are signed in, the command will stop hanging.

**Comment by helfi92** [comment]
> So this pagination is needed only for the roles page? 

It will be needed for:
1. https://taskcluster-web.netlify.com/auth/roles
2. https://taskcluster-web.netlify.com/auth/scopes (roles tab only)

> Should I do it in DataTable component so it will apply to all

We have two separate components for tables. `DataTable` is for displaying a table without pagination whereas `ConnectionDataTable` is for tables with pagination. For the roles table, you will want to use the `ConnectionDataTable` since we want a table with pagination. Feel free to start with (1) then once that's merged, you can create another PR for (2).

**Comment by arshadkazmi42** [comment]
I was exploring the existing pages which are using `ConnectionDataTable`
Looks like those queries are using `edges` and `nodes` and `roles` query does not implements `edges` and `nodes`
Is there some new query implemented for this at backend?

**Comment by arshadkazmi42** [comment]
Found it here https://github.com/taskcluster/taskcluster-web-server/pull/71#issuecomment-449107612

**Comment by arshadkazmi42** [comment]
@helfi92 Created a PR here
https://github.com/taskcluster/taskcluster-web/pull/337
So I have created pagination, but on clicking next list is not getting refreshed.
Can you check it once, if I have missed something

**Comment by arshadkazmi42** [comment]
Will start working on it. :+1:

**Comment by arshadkazmi42** [comment]
@helfi92 So I was thinking, should we just use 
https://github.com/taskcluster/taskcluster-web/blob/master/src/views/Roles/ViewRoles/Roles.jsx
Like its used here
https://github.com/taskcluster/taskcluster-web/blob/master/src/views/Roles/ViewRoles/index.jsx#L45

Or separate implementation inside `View/Scopes` will be better approach for this?

**Comment by helfi92** [comment]
It probably makes sense to use `ViewRoles/Roles.jsx`. Feel free to use it.

**Comment by arshadkazmi42** [comment]
I explored it a bit, RoleScope is using RolesScopesTable, which is a bit different from RolesTable.
I think using `ViewRoles/Roles.jsx` will not work.
I think I should implement it separately

**Comment by helfi92** [comment]
Sounds good.

**Comment by arshadkazmi42** [comment]
I started working on it and found that ScopeRoles requires,
`expandedScopes`, `scopes` fields to be accessed in listRolesId.
Is it not exposed?

**Comment by helfi92** [comment]
That's a good question. @djmitche Would it be okay if we add `expandedScopes` and `scopes` to the new listRoleIds endpoint? We will need that to successfully paginate the `auth/scopes` view (roles table).

**Comment by helfi92** [comment]
Good question. Let's ask dustin in https://bugzilla.mozilla.org/show_bug.cgi?id=1455130#c16.

**Comment by djmitche** [comment]
I'm not sure what the problem is?  Is GraphQL incapable of returning a list of identifiers without all of their data included?

**Comment by helfi92** [comment]
We can use the `filter`parameter to filter on the list of role Ids. I'm not sure however how this relates to the issue. Are you saying we are able to get the `expandedScopes` by using a filter on the `listRolesId` query?

**Comment by djmitche** [comment]
No, I'm saying if you want all of that information, you should just call the `listRoles` method that has existed since forever.

**Comment by helfi92** [comment]
I spoke with @djmitche and we concluded that we need a new method in [taskcluster-auth](https://github.com/taskcluster/taskcluster-auth), namely `listRoles2`. I created https://bugzilla.mozilla.org/show_bug.cgi?id=1517494. Is this something you are interested in tackling? We would be happy to help. If you are, please comment directly on the bugzilla bug to claim it.

Regarding https://taskcluster-web.netlify.com/auth/scopes, we think that we should not go forward with the pagination just yet. We will tackle this issue later when we have a better idea how we want to display scopes to the user.

Thank you again for integrating pagination in `/auth/roles`. This is a huge win in terms of UX :)

**Comment by arshadkazmi42** [comment]
@helfi92 Thanks for the figuring that out.
I will take up the bug for `listRoles2`.

**Comment by helfi92** [comment]
My pleasure! Thanks for your help as always.
